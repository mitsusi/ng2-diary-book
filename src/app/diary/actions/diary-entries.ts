import { Action } from '@ngrx/store';

import { DiaryEntry } from 'ng2-diary-book-shared-models';

export const LOAD_LIST = '[Diary entries] Load list';
export const LOAD_LIST_SUCCESS = '[Diary entries] Load list success';
export const LOAD_LIST_FAILURE = '[Diary entries] Load list failure';

export const CREATE_ENTRY = '[Diary entries] Create entry';
export const CREATE_ENTRY_SUCCESS = '[Diary entries] Create entry success';
export const CREATE_ENTRY_FAILURE = '[Diary entries] Create entry failure';

export const EDIT_ENTRY = '[Diary entries] Edit entry';
export const EDIT_ENTRY_SUCCESS = '[Diary entries] Edit entry success';
export const EDIT_ENTRY_FAILURE = '[Diary entries] Edit entry failure';

export const GET_EDIT_ENTRY = '[Diary entries] Get edit entry';
export const GET_EDIT_ENTRY_SUCCESS = '[Diary entries] Get edit entry success';
export const GET_EDIT_ENTRY_FAILURE = '[Diary entries] Get edit entry failure';
export const GET_EDIT_ENTRY_CLEAR = '[Diary entries] Get edit entry clear';

/**
 * Every action is comprised of at least a type and an optional
 * payload. Expressing actions as classes enables powerful
 * type checking in reducer functions.
 *
 * See Discriminated Unions: https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions
 */
export class LoadListAction implements Action {
  readonly type = LOAD_LIST;

  constructor(public payload?: string[]) {}
}

export class LoadListSuccessAction implements Action {
  readonly type = LOAD_LIST_SUCCESS;

  constructor(public payload: any[]) {}
}

export class LoadListFailureAction implements Action {
  readonly type = LOAD_LIST_FAILURE;

  constructor(public payload: any) {}
}

export class CreateEntryAction implements Action {
  readonly type = CREATE_ENTRY;

  constructor(public payload: DiaryEntry) {}
}

export class CreateEntrySuccessAction implements Action {
  readonly type = CREATE_ENTRY_SUCCESS;

  constructor(public payload: DiaryEntry) {}
}

export class CreateEntryFailureAction implements Action {
  readonly type = CREATE_ENTRY_FAILURE;

  constructor(public payload: any) {}
}

export class EditEntryAction implements Action {
  readonly type = EDIT_ENTRY;

  constructor(public payload: DiaryEntry) {}
}

export class EditEntrySuccessAction implements Action {
  readonly type = EDIT_ENTRY_SUCCESS;

  constructor(public payload: DiaryEntry) {}
}

export class EditEntryFailureAction implements Action {
  readonly type = EDIT_ENTRY_FAILURE;

  constructor(public payload?: any) {}
}

export class GetEditEntryAction implements Action {
  readonly type = GET_EDIT_ENTRY;

  constructor(public payload: string) {}
}

export class GetEditEntrySuccessAction implements Action {
  readonly type = GET_EDIT_ENTRY_SUCCESS;

  constructor(public payload: DiaryEntry) {}
}

export class GetEditEntryFailureAction implements Action {
  readonly type = GET_EDIT_ENTRY_FAILURE;

  constructor(public payload?: any) {}
}

export class GetEditEntryClearAction implements Action {
  readonly type = GET_EDIT_ENTRY_CLEAR;

  constructor() {}
}

/**
 * Export a type alias of all actions in this action group
 * so that reducers can easily compose action types
 */
export type Actions =
  | LoadListAction
  | LoadListSuccessAction
  | LoadListFailureAction
  | CreateEntryFailureAction
  | CreateEntrySuccessAction
  | CreateEntryFailureAction
  | EditEntryFailureAction
  | EditEntrySuccessAction
  | EditEntryAction
  | GetEditEntryAction
  | GetEditEntrySuccessAction
  | GetEditEntryClearAction
  | GetEditEntryFailureAction;
