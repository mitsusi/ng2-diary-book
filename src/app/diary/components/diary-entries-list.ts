import {
  Component,
  EventEmitter,
  Input,
  Output,
  ChangeDetectionStrategy,
} from '@angular/core';

import { DiaryEntry } from 'ng2-diary-book-shared-models';

@Component({
  selector: 'diary-entries-list',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <div>
      <div *ngFor="let entry of entries">
        <diary-entries-list-item
          [entry]="entry"
          [isEditDisabled]="isEditDisabled"
          (onClick)="onClick.emit($event)"
        ></diary-entries-list-item>
      </div>
    </div>
  `,
  styles: [
    `diary-entries-list-item:not(:first-child) {
      padding: 20px;
    }`,
  ],
})
export class EntryListComponent {
  @Input() entries: DiaryEntry[];
  @Input() isEditDisabled = false;
  @Output() onClick = new EventEmitter();

  constructor() {}
}
