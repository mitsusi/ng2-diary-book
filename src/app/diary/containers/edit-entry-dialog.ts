import { Component, Inject } from '@angular/core';
import { MAT_DIALOG_DATA } from '@angular/material';
import { DiaryEntry } from 'ng2-diary-book-shared-models';

@Component({
  selector: 'diary-edit-dialog',
  template: `
    <diary-entry-edit
      [entry]="entry"
    ></diary-entry-edit>
  `,
})
export class EntryEditDialogComponent {
  entry: DiaryEntry;

  constructor(@Inject(MAT_DIALOG_DATA) public data: any) {
    this.entry = data;
  }
}
