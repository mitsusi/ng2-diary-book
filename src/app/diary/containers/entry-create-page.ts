import { Component } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';

@Component({
  template: `
    <mat-card>
      <mat-card-header>
        <h4>Добавление записи</h4><span (click)="goBack();">Назад <mat-icon><i class="material-icons">arrow_back</i></mat-icon></span>
      </mat-card-header>
      <diary-entry-create></diary-entry-create>
    </mat-card>
  `,
})
export class EntryCreateContainer {
  goBack() {
    this.router.navigate(['..'], { relativeTo: this.route });
  }

  constructor(private router: Router, private route: ActivatedRoute) {}
}
