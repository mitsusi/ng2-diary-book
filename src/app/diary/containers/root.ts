import { Component } from '@angular/core';

@Component({
  template: `
    <div class="root">
      <router-outlet></router-outlet>
    </div>
  `,
  styleUrls: [`../../core/styles/general.styles.css`],
})
export class DiaryRootComponent {}
