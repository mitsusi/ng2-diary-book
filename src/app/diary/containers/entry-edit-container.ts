import {
  Component,
  Input,
  ChangeDetectionStrategy,
  Output,
  EventEmitter,
} from '@angular/core';
import { Store } from '@ngrx/store';

import { DiaryEntry } from '../../shared/models/diary-entry';

import * as fromRoot from '../reducers';
import * as diaryActions from '../actions/diary-entries';

@Component({
  selector: 'diary-entry-edit',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <diary-entry-form
      [entry]="entry"
      [options]="entryDetailComponentOptions"
      (submit)="onSubmit($event)"
    ></diary-entry-form>
  `,
})
export class EntryEditContainerComponent {
  @Input() entry: DiaryEntry;
  entryDetailComponentOptions: object;

  constructor(private store: Store<fromRoot.State>) {
    this.entryDetailComponentOptions = {
      buttonLabel: 'Edit',
    };
  }

  onSubmit(entryToEdit: DiaryEntry) {
    this.store.dispatch(new diaryActions.EditEntryAction(entryToEdit));
  }
}
