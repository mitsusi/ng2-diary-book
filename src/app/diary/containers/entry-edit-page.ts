import { Component, OnInit } from '@angular/core';
import { ActivatedRoute, ParamMap, Router } from '@angular/router';
import { Observable } from 'rxjs/Observable';
import { DiaryEntry } from 'ng2-diary-book-shared-models';
import { Store } from '@ngrx/store';

import * as diaryActions from '../actions/diary-entries';
import * as routerActions from '../../core/actions/router';
import * as fromDiary from '../reducers';
import * as fromRoot from '../../reducers';

@Component({
  template: `
    <md-card>
      <md-card-header>
        <h4>Изменение записи</h4><span (click)="goBack();">Назад <md-icon><i class="material-icons">arrow_back</i></md-icon></span>
      </md-card-header>
      <diary-entry-edit
        [entry]="entryToEdit$ | async"
      ></diary-entry-edit>
    </md-card>
  `,
})
export class EntryEditContainer {
  entryToEdit$: Observable<DiaryEntry>;

  goBack() {
    this.store.dispatch(
      new routerActions.Go({
        path: ['../..'],
        extras: { relativeTo: this.route },
      }),
    );
  }

  constructor(
    private router: Router,
    private route: ActivatedRoute,
    private store: Store<fromDiary.State>,
  ) {
    store
      .select(fromRoot.getUserIsLoggedIn)
      .filter(isUserLoggedIn => isUserLoggedIn)
      .subscribe(() => {
        this.route.paramMap.subscribe((params: ParamMap) =>
          this.store.dispatch(
            new diaryActions.GetEditEntryAction(params.get('key')),
          ),
        );
      });
    this.entryToEdit$ = this.store.select(fromDiary.getDiaryToEditEntry);
  }
}
