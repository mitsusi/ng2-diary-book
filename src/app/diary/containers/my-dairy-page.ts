import {
  Component,
  ChangeDetectionStrategy,
  ChangeDetectorRef,
} from '@angular/core';
import { Observable } from 'rxjs/Observable';
import { Store } from '@ngrx/store';

import * as fromRoot from '../../reducers';
import * as fromDiary from '../reducers';

import * as diaryEntries from '../actions/diary-entries';

import { DiaryEntry } from '../../shared/models/diary-entry';

import { ActivatedRoute, Router } from '@angular/router';

@Component({
  selector: 'diary-page',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
    <common-show-if-logged-in>
      <diary-entries-list
        [entries]="diaryEntries$ | async"
        (onClick)="goToEdit($event)"
      ></diary-entries-list>
      <button md-mini-fab class="example-fab" (click)="goToAdd()">
        <md-icon>add</md-icon>
      </button>
    </common-show-if-logged-in>
  `,
  styles: [
    `diary-page {
      position: absolute;
      width: 100%;
      height: 100%;
    }`,
    `.example-fab {
      position: fixed;
      right: 20px;
      bottom: 10px;
    }`,
  ],
})
export class MyDairyPageComponent {
  diaryEntries$: Observable<DiaryEntry[]>;

  constructor(
    private store: Store<fromDiary.State>,
    private changeDetectorRef: ChangeDetectorRef,
    private route: ActivatedRoute,
    private router: Router,
  ) {
    this.diaryEntries$ = store.select(fromDiary.getDiaryEntries);
    this.diaryEntries$.subscribe(data => console.log(data));
    store.select(fromRoot.getUserIsLoggedIn).subscribe(isLoggedIn => {
      if (isLoggedIn) {
        this.store.dispatch(new diaryEntries.LoadListAction());
        changeDetectorRef.markForCheck();
      }
    });
  }

  goToAdd() {
    this.router.navigate(['./add'], { relativeTo: this.route });
  }

  goToEdit(entryToEdit: DiaryEntry) {
    this.router.navigate(['./edit', entryToEdit.$key], {
      relativeTo: this.route,
    });
  }
}
