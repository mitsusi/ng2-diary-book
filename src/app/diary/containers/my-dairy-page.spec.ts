/* tslint:disable:no-unused-variable */
import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { Store, StoreModule, combineReducers } from '@ngrx/store';

import * as fromFeature from '../reducers';
import * as fromRoot from '../../reducers';

import { MyDairyPageComponent } from './my-dairy-page';

import * as userActions from '../../core/actions/user';
import * as diaryActions from '../actions/diary-entries';

import { EntryListMockMetadata } from '../../core/components/__mocks__/entry-list-mock-metadata';
import { EntryListComponent } from '../components/diary-entries-list';

import { ComponentsModule as DiaryComponentsModule } from '../components/index';
import { MdIconModule } from '@angular/material';
import { CommonShowIfLoggedInComponent } from '../../core/containers/common-show-if-logged-in';
import { NotLoggedInComponent } from '../../core/components/not-logged-in';

describe('MyDairyPageComponent', () => {
  let component: MyDairyPageComponent;
  let fixture: ComponentFixture<MyDairyPageComponent>;
  let store: Store<fromFeature.State>;

  beforeEach(
    async(() => {
      TestBed.overrideComponent(EntryListComponent, {
        set: EntryListMockMetadata,
      })
        .configureTestingModule({
          imports: [
            MdIconModule,
            DiaryComponentsModule,
            StoreModule.forRoot({
              ...fromRoot.reducers,
              diary: combineReducers(fromFeature.reducers),
            }),
          ],
          declarations: [
            MyDairyPageComponent,
            CommonShowIfLoggedInComponent,
            NotLoggedInComponent,
          ],
          providers: [],
        })
        .compileComponents();

      store = TestBed.get(Store);
    }),
  );

  beforeEach(() => {
    fixture = TestBed.createComponent(MyDairyPageComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('полная имитация', () => {
    it('', () => {
      const userDataMock = {
        /*...*/
      };
      store.dispatch(new userActions.LoadUserAction(userDataMock));
      store.dispatch(
        new diaryActions.LoadListSuccessAction([
          {
            createDate: '2017-08-19T23:22:34.856Z',
            date: '2017-08-19T21:00:00.000Z',
            message: '13123123asdasdasd',
            $key: '-KrwnXF5o7VRJDTXJw-I',
          },
          {
            createDate: '2017-08-17T15:24:17.399Z',
            date: '2017-08-16T21:00:00.000Z',
            message: '123123123123123',
            $key: '-KrkmsupthRsqGXqVBeK',
          },
        ]),
      );

      fixture.detectChanges();

      expect(fixture).toMatchSnapshot();
    });
  });
});
