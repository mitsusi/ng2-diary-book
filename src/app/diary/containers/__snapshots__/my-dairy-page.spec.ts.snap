// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyDairyPageComponent should match snapshot 1`] = `
<diary-page
  changeDetectorRef={[Function ViewRef_]}
  dialogFactory={[Function DialogFactoryService]}
  diaryEntries$={[Function Store]}
  store={[Function Store]}
>
  <common-show-if-logged-in
    _ngcontent-c3=""
  >
    <div
      ng-reflect-ng-switch="false"
    >
      <div>
        <diary-not-logged-in>
          You are not logged in. Click to log in.
        </diary-not-logged-in>
      </div>
    </div>
  </common-show-if-logged-in>
</diary-page>
`;

exports[`MyDairyPageComponent полная имитация  1`] = `
<diary-page
  changeDetectorRef={[Function ViewRef_]}
  dialogFactory={[Function DialogFactoryService]}
  diaryEntries$={[Function Store]}
  store={[Function Store]}
>
  <common-show-if-logged-in
    _ngcontent-c6=""
  >
    <div
      ng-reflect-ng-switch="true"
    >
      <div>
        <diary-entry-list
          _ngcontent-c6=""
          _nghost-c7=""
          ng-reflect-entries="[object Object],[object Object"
        >
          entry-list
        </diary-entry-list>
        <button
          _ngcontent-c6=""
          class="example-fab"
          md-mini-fab=""
        >
          <md-icon
            _ngcontent-c6=""
            class="mat-icon material-icons"
            role="img"
            aria-hidden="true"
          >
            add
          </md-icon>
        </button>
      </div>
    </div>
  </common-show-if-logged-in>
</diary-page>
`;
